File: combination-sum/combination-sum.cs
public class Solution {
    List<IList<int>> result;
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        result = new List<IList<int>>();
        
        if(candidates.Length ==0) return result;
        
        helper(candidates,target,0,0,new List<int>());
        return result;
        
    }
    
    private void helper(int[] candidates,int target,int index,int currsum,List<int> temp)
    {
       if(currsum > target ||index>=candidates.Length)
       {
           return;
       }
        if(currsum == target)
        {
            result.Add(new List<int>(temp));
            return;
        }
        
        helper(candidates,target,index+1,currsum,temp);
        temp.Add(candidates[index]);
        helper(candidates,target,index,currsum + candidates[index],temp);
        for(int j=0;j<temp.Count;j++)
        {
            Console.WriteLine(temp[j]);
        }
        temp.Remove(candidates[index]);
    }
}

File: assignments/src/Combination_Sum.java
import java.util.*;

public class Combination_Sum {

    // Approach: Backtracking using choose or don't choose as the logic
    // Time: O(N^((T/M) + 1)) where N = no. of candidates, T = target value, M = minimal value among candidates
    // Space: O(T/M)
    List<List<Integer>> result;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        result = new ArrayList();
        helper(candidates, 0, new ArrayList(), target);
        return result;
    }

    private void helper(int[] candidates, int i, List<Integer> path, int target) {
        // base case
        if (target == 0) {
            result.add(new ArrayList(path));
            return;
        }

        if (target < 0 || i == candidates.length) return;

        // logic
        // don't choose
        helper(candidates, i+1, path, target);
        // choose
        // action
        path.add(candidates[i]);
        // recurse
        helper(candidates, i, path, target - candidates[i]);
        // backtrack
        path.remove(path.size() - 1);
    }
}